Display warning about spaces in underlying values for multiple selection questions. Fixes #1964

bartekfi opened the pull request:
Closes  #1964 

#### What has been done to verify that this works as intended?
Tested with [this form](https://www.dropbox.com/s/ebs4uub4k4qiaxd/multi-warning.xml?dl=0)

<img src="https://user-images.githubusercontent.com/721925/40590898-8ab563b8-6207-11e8-8ae0-49447c823e82.png" width="180" height="320"> <img src="https://user-images.githubusercontent.com/721925/40590899-8adb87f0-6207-11e8-9650-ce2d45a6b2a3.png" width="180" height="320"> <img src="https://user-images.githubusercontent.com/721925/40590900-8b0e19cc-6207-11e8-82ee-b89506d91319.png" width="180" height="320"> <img src="https://user-images.githubusercontent.com/721925/40590901-8b2a86fc-6207-11e8-9ea8-a49f0dca0e2f.png" width="180" height="320"> <img src="https://user-images.githubusercontent.com/721925/40590902-8b48c9fa-6207-11e8-872c-4311234db47c.png" width="180" height="320"> <img src="https://user-images.githubusercontent.com/721925/40590903-8b6575a0-6207-11e8-8cd3-051b7248a8cd.png" width="180" height="320">


#### Why is this the best possible solution? Were any other approaches considered?
The solution generalises solution from https://github.com/opendatakit/collect/pull/2024
It adds a TextView in help text layout, and uses it to display a warning.

#### Are there any risks to merging this code? If so, what are they?
The solution can be used to displayed any warning, but is not good for displaying multiple unrelated warnings, as it will always replace previously set warning with a new one.

Commit:
Extract detecting and rendering warning of spaces in underlying values into a class

Commit:
Extend QuestionWidget to allow generic showing of warning messages

Additionally add tests for SpacesInUnderlyingValuesWarning.

Additional consequence is chaining how the warning text is rendered.
Instead of in code padding, the text is now aligned the to the help text widget.

Commit:
Show warning about spaces in underlying values in GridMultiWidget

Commit:
Show warning about spaces in underlying values in SpinnerMultiWidget

Commit:
Show warning about spaces in underlying values in MultiAutocompleteWidget

Commit:
Show warning about spaces in underlying values in SelectMutliImageMapWidget

Commit:
Show warning about spaces in underlying values in LabelWidget

Commit:
Clean up: remove RelativeLayout leftover qualifier from LinearLayout

It does nothing in LinearLayout

bartekfi commented:
Unrelated to this PR, I've noticed that LabelWidget layout is broken (not affected by this PR).
I also think that enumerating underlying values might not provide any value to app user. 

Instead I would rephrase the warning into something like "Warning: values in this question contain spaces - it might not be saved correctly."

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java:
```diff
@@ -0,0 +1,116 @@
+
+import java.util.List;
+
+public class SpacesInUnderlyingValuesWarning {
```
A doc comment here would help readers understand what this is

bartekfi replied:
@dcbriccetti Added documentation in c4238ea

Commit:
Add documentation for SpacesInUnderlyingValuesWarning

This should be merged into "Extend QuestionWidget to allow generic showing of warning messages" after the review process is complete.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java:
```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+    interface SpacesInUnderlyingValues {
```
It feels too bad that the interface name is so specific. This really could be called something more generic like `UnderlyingValueChecker`, right? I also don't love using `Impl` as part of a class name so perhaps that could then become `SpacesUnderlyingValueChecker` or something like that.

bartekfi replied:
Yes, that looks much better, I'll apply the changes.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java:
```diff
@@ -0,0 +1,142 @@
+        String create(List<SelectChoice> invalidValues, Context context);
+    }
+
+    private static class WarningTextCreatorImpl implements WarningTextCreator {
```
Instead of `WarningTextCreatorImpl` how about something like `SpacesWarningTextCreator`?

lognaturel reviewed (commented):
Thank you, @bartoszfilipowicz! I made a couple of small naming suggestions inline.

lognaturel commented:
> I've noticed that LabelWidget layout is broken

Do you mean in the sense that hint or warning text may overlap with labels? This is actually ok. In actual usage it's rare to have a hint there and it does look good on larger screens where that type of layout is more commonly used. If it's something else, please do file an issue!

> I also think that enumerating underlying values might not provide any value to app user.

I see your point. What I like about including the value(s) is that at least it gives the user something specific they can search for in their form. Users don't always have deep knowledge of form structure and so they may not know what "values in this question" refers to.

bartekfi commented:
@lognaturel  PR updated with propsed name changes. 

> Users don't always have deep knowledge of form structure and so they may not know what "values in this question" refers to.

Understood. Seems like this information is mostly actionable for the form creator, who's testing the form, rather than the person who actually collects the data.

I think this kind of error requires stronger accent, as effectively it makes the data collected for the entire question invalid or incomplete. I propose using red warning colour as in the screenshot below. WDYT @lognaturel?

<img src="https://user-images.githubusercontent.com/721925/40775756-1ce43ef8-64c9-11e8-873f-ed2dc4db0014.png" width="180" height="320">

lognaturel commented:
> Seems like this information is mostly actionable for the form creator, who's testing the form, rather than the person who actually collects the data.

Exactly!

> I propose using red warning colour as in the screenshot below. 

Thanks, great idea. Once you make that change we're ready for QA unless @dcbriccetti has any other suggestions or questions.

dcbriccetti commented:
>unless @dcbriccetti has any other suggestions or questions.

Nope

Commit:
Render warning on question widget in red

Use new attr ?warningColor to make sure warning color is controlled in
dark and light themes. In this case the color is the same for both white
and dark themes.

bartekfi commented:
Added one more requested change: changed color of the warning to red. It looks as follows in light and dark themes:

<img src="https://user-images.githubusercontent.com/721925/41056231-c414c52c-69c3-11e8-9b2f-ceee3f2ff21d.png" width="180" height="320"> <img src="https://user-images.githubusercontent.com/721925/41056251-d095ec04-69c3-11e8-8e4a-4e6a659d1bd4.png" width="180" height="320">

This PR is ready for testing!

grzesiek2010 commented on collect_app/src/main/res/layout/help_text_layout.xml:
```diff
@@ -53,5 +53,6 @@
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
             android:visibility="gone"
+            android:textColor="?warningColor"
```
If it's red in both themes it doesn't really make sense to do that in that way, why don't you just use `@color/red` here?

bartekfi replied:
@grzesiek2010 I see your point. My thinking in this matter is as follows:

1. The fact we can use the same color (`@color/red`) for warning in both themes is a coincidence and not a rule or design decision
2. Using attribute allows us to name the color according to its use, rather than value "red". So using `?warningColor` is a nice way to create an alias that is defined by usage not value. (alternatively we can create a global color such as `@color/warningColor`
3. Warning color depends on the theme. It depends on the same looks of theme as `?primaryTextColor`, `?colorAccent`. So theme is the context of the color, and the reason it has this value, so it seems like a natural place to put it. 

WDYT?

bartekfi replied:
Using attributes over direct color reference is also recommended by ODK's contribution guidelines.

lognaturel replied:
This is not one I feel too strongly about but in general I do prefer semantic naming so I like the approach @bartoszfilipowicz has taken. @grzesiek2010 do you find his arguments convincing?

grzesiek2010 replied:
> do you find his arguments convincing?

not entirely but I'm not going to complain since it's a small thing. I just noticed the same case is with `colorAccent`, both colors are the same across both themes so adding them to both themes is a duplication I think but fair enough.

mmarciniak90 commented:
@bartoszfilipowicz @lognaturel 
The toast was added to `Multi-select` widget as a part of #2024.
Other widgets do not contain toast, just warning. I feel that we should take care of behavior compatibility for all `multi` widgets.
We need to decide whether add toast for others widgets or remove it from one. 
Removing refers to only one widget so should be faster but on the other hand, the user can be confused what option can cause a problem in the case with images.
![screenshot_20180608-152035](https://user-images.githubusercontent.com/3907554/41157892-8e884bbe-6b16-11e8-826e-3ddc012250c9.png)

mmarciniak90 commented:
@opendatakit-bot label "needs discussion"

lognaturel commented:
Thanks, @mmarciniak90! I don't think the toast adds much information. Like @bartoszfilipowicz stated previously, the warning is for a form developer rather than for someone collecting data so they should be able to figure out where the changes are needed.

I think we should merge this and then file another issue to remove the toast.

mmarciniak90 commented:
@lognaturel sounds good to me! I created a new issue: #2284

**Tested with success**

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

bartekfi commented:
Thanks for testing and for reporting issue @mmarciniak90 👍 
@lognaturel I'll post a PR with fix to #2284 as soon as this is merged.
